File: HW#02/src/Unal_Baris.dart
void main() {

  // PART 1: Should the conditional mechanism be an integral part of the exit (conditional or unconditional exit)?

  // Both conditional and unconditional exits are possible to implement in Dart. A simple conditional exit is as follows.
  print( "\nCONDITIONAL EXIT \nCountdown from 10:" );
  var x = 9;
  while( x >= 0 ){
    print(x);
    x--;
  }

  // However, conditional mechanism is NOT an integral part of the exit in Dart. One can use "break" statement. There are two types of break statements. First one is the unlabeled one.
   print( "\nUNCONDITIONAL UNLABELED EXIT \nCounting up to 10:" );
  for( var i = 0; i < 10; i++ ) {
    if( i == 5 ) {
      break;
    }
    print(i);
  }
  // The loop is exited once i becomes 5, which causes to stop counting at after 4.

  
  // PART 2: Should only one loop body be exited or can enclosing loops also be exited (labeled or unlabeled exit)?

  // The second option to use the "break" statement is to state a label which points a line of code to continue executing.
  print( "\nUNCONDITIONAL LABELED EXIT \nThree nested loops with break statement: " );
  for( var i = 0; i <= 2; i++ ) {
    print( "Inside first loop i = " + i.toString() );
    loop1:
    for( var j = 0; j <= 2; j++ ) {
      print( "Inside second loop j = " + j.toString() );
      for( var k = 0; k <= 2; k++ ) {
        if( j > 1 ) {
          break loop1;
          // exits the second loop and continues execution of the first loop's body
        }
        print( "Inside third loop k = " + k.toString());
      }
    }
  }
  // When we reach the innermost loop when j==2 i.e, for the third time, the "break" statement is executed and the block which label "loop1" points to i.e, the second loop, is exited. 


    // The "continue" keyword can be used with or without labels. The one without a label is as follows.
  print( "\nUNLABELED CONTINUE \nTwo nested loops with unlabeled continue statement:" );
  for( var i = 0; i <= 3; i++ ) {
    print( "Inside first loop i = " + i.toString() );
    for( var j = 0; j <= 3; j++ ) {
      if( j == 1 ){
        continue;
      }
      print( "Inside second loop j = " + j.toString() );
    }
  }
  // When j==1, continue statement is executed and the program continues to check the condition for the innermost loop of the body that "continue" statement is in and executes the body if the condition holds.

  // A "continue" statement with a label is as follows.
  print( "\nLABELED CONTINUE \nTwo nested loops with labeled continue statement:" );
  loopOuter:
  for( var i = 0; i <= 3; i++ ) {
    print( "Inside first loop i = " + i.toString() );
    for( var j = 0; j <= 3; j++ ) {
      if( j == 1 ){
        continue loopOuter;
      }
      print( "Inside second loop j = " + j.toString() );
    }
  }
  // When j==1, continue statement is executed and the program continues to check the condition for the outer loop since the label "loopOuter" points to that loop. Therefore, we never see j become 1, 2 or 3.
  
}

File: HW#02/src/Unal_Baris.lua

print("Hello, world!")

-- PART 1: Should the conditional mechanism be an integral part of the exit (conditional or unconditional exit)?

-- Both conditional and unconditional exits are possible to implement in Lua. A simple conditional exit is as follows.
print( "\nCONDITIONAL EXIT \nCountdown from 10:" )
x = 9;
while( x >= 0 )
do
  print( x );
  x = x - 1;
end

-- However, conditional mechanism is NOT an integral part of the exit in Lua. One can use "break" statement for unconditional unlabeled exit.
print( "\nUNCONDITIONAL UNLABELED EXIT BREAK \nCounting up to 10 interrupted by break:" );
for i = 0, 10, 1 
do
  if i == 5 then break end
  print(i);
end
-- The loop is exited once i becomes 5, which causes to stop counting at after 4.

-- PART 2: Should only one loop body be exited or can enclosing loops also be exited (labeled or unlabeled exit)?
-- The "break" statement only exits the innermost loop in Lua.
print( "\nUNCONDITIONAL UNLABELED EXIT \nThree nested loops with break statement: " );
for i = 0, 2, 1 
do
  print( "\tIn the first loop i = ", i );
  for j = 0, 2, 1
  do 
    print( "\t\tIn the second loop j = ", j );
    for k = 0, 2, 1
    do 
      print( "\t\t\tIn the third loop k = ", k );
      if i == 1 then break end
    end
  end 
end
-- Only the third loop was exited when i equals 1.

print( "\nUNCONDITIONAL LABELED EXIT \nNested for loops with a goto statement to illustrate which statement is executed, i.e., which for loop is being exited: " );
x = 0;
for i = 0, 10, 1 do
  for j = 0, 10, 1 do
    goto out
  end
  x = 1;
end
::out::
print( "x = ", x);
-- In the inner loop, the program jumps to the line of the label 'out' and therefore x never becomes 1.

-- In Lua, one can only exit the innermost loop body with "break" statement. There is no way to exit outer loops with "break" statement because there is no labeled "break" statements unlike JavaScript and Dart.
print( "\nUNCONDITIONAL LABELED EXIT WITH GOTO \nThree nested loops with goto statement: " );
i = 0;
while i <= 2 
do
  i = i + 1;
  print( "In the first loop i = ", i );
  j = 0;
  ::exit::
  while j <= 2
  do 
    j = j + 1;
    print( "\tIn the second loop j = ", j );
    k = 0;
    while k <= 2
    do 
      k = k + 1;
      print( "\t\tIn the third loop k = ", k );
      if i == 2 then 
          goto exit 
        end
    end
  end 
end





 